torch/optim/optimizer.py

```diff
@@ -168,7 +168,8 @@ def zero_grad(self):
         for group in self.param_groups:
             for p in group['params']:
                 if p.grad is not None:
-                    p.grad.detach_()
+                    if p.grad.grad_fn is not None:
+                        p.grad.detach_()
```

Comment:
On top of removing `.grad_fn`, it also removes the `.requires_grad` flag. I think you want to add a `p.grad.requires_grad_(False)` even when there is no grad_fn.